### PR TITLE
feat(learn): show leave page confirmation if project code not saved

### LINKED ISSUE #46810: Camper Feedback on Code Saving UX
### Describe the Issue

Received this feedback via our support email:

> FYI, the "Save Your Code" button is not visible on the screen because the challenge description pushes it below the scroll window. I didn't even know that button existed until I scrolled down to the bottom of the window. This is a terrible UI, especially in a world where most online tools simply autosave.

These certification projects take a significant amount of work, so losing progress can be remarkably frustrating. We need to reconsider the logical flow we have here and see how we can improve it to prevent this going forward.

### Affected Page

https://www.freecodecamp.org/learn/2022/responsive-web-design/build-a-tribute-page-project/build-a-tribute-page

### Steps to Reproduce

1. Visit the above page
2. See that the save code button is not in view 

### Expected behavior

We should evaluate the location of that button - it might be better up in the top action row?

Additionally, we should revisit our previous discussion around auto saving.

### Screenshots

image

### System

- Device: [e.g. iPhone 6, Laptop]
- OS: [e.g. iOS 14, Windows 10, Ubuntu 20.04]
- Browser: [e.g. Chrome, Safari]
- Version: [e.g. 22]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kamalsinghb**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #46810 

[Screencast from 2024-10-21 13-39-31.webm](https://github.com/user-attachments/assets/d15935a8-a738-45c5-814f-4dc505b8335b)

### CODE REVIEW

**Review by Sembauke — CHANGES_REQUESTED** [review]
Hey @kamalxdev,
Could we do something like this? It is a little bit cleaner.

```js
    const hasSaved = savedChallenges[0]?.challengeFiles;

    for (let index = 0; index < challengeFiles.length; index++) {
      if (!hasSaved) isFilesSaved = false;

      const hasChangedContentAfterSave =
        savedChallenges[0]?.challengeFiles?.[index]?.contents !=
        challengeFiles?.[index]?.contents;

      if (hasChangedContentAfterSave) isFilesSaved = false;
    }
```

**Comment by kamalxdev** [comment]
> Hey @kamalxdev, Could we do something like this? It is a little bit cleaner.
> 
> ```js
>     const hasSaved = savedChallenges[0]?.challengeFiles;
> 
>     for (let index = 0; index < challengeFiles.length; index++) {
>       if (!hasSaved) isFilesSaved = false;
> 
>       const hasChangedContentAfterSave =
>         savedChallenges[0]?.challengeFiles?.[index]?.contents !=
>         challengeFiles?.[index]?.contents;
> 
>       if (hasChangedContentAfterSave) isFilesSaved = false;
>     }
> ```

You mean like this: 
`
  useEffect(() => {
    let isFilesSaved = true;

    const savedChallengeFiles=savedChallenges[0]?.challengeFiles;

    if (!challengeFiles) return;
    
    for (let index = 0; index < challengeFiles.length; index++) {
      let hasChangedContentAfterSave = savedChallengeFiles && savedChallengeFiles[index]?.contents !=challengeFiles[index]?.contents;

      if (!savedChallengeFiles || hasChangedContentAfterSave) {
        isFilesSaved = false;
      }
    }

    if (isFilesSaved) return;

    function onbeforeUnload(e: BeforeUnloadEvent) {
      e.preventDefault();
    }

    window.addEventListener('beforeunload', onbeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onbeforeUnload);
    };
  }, [challengeFiles, savedChallenges]);
`

**Comment by Sembauke** [comment]
Sure, just make sure `hasChangedContentAfterSave` is a constant!

**Comment by kamalxdev** [comment]
> Sure, just make sure `hasChangedContentAfterSave` is a constant!

okay

**Comment by kamalxdev** [comment]
@Sembauke  changes updated

File: client/src/templates/Challenges/classic/show.tsx [hunk 1]
```diff
@@ -301,6 +301,38 @@ function ShowClassic({
     setUsingKeyboardInTablist(usingKeyboardInTablist);
   };
 
+  // leave page confirmation if code is not saved
+
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/classic/show.tsx:305]:
```suggestion
  // Show leave page confirmation if code is not saved
```
Nit

File: client/src/templates/Challenges/classic/show.tsx [hunk 2]
```diff
@@ -301,6 +301,38 @@ function ShowClassic({
...
+    if (isFilesSaved) return;
+
+    function onbeforeUnload(e: BeforeUnloadEvent) {
+      e.preventDefault();
+    }
+
+    window.addEventListener('beforeunload', onbeforeUnload);
+
+    return () => {
+      window.removeEventListener('beforeunload', onbeforeUnload);
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/classic/show.tsx:332]:
```suggestion
    const savedChallengeFiles = savedChallenge?.challengeFiles;
    const savedContents = savedChallengeFiles?.map(file => file.contents);
    const currentContents = challengeFiles?.map(file => file.contents);

    const isSaved =
      savedChallengeFiles && isEqual(savedContents, currentContents);

    function handleBeforeUnload(e: BeforeUnloadEvent) {
      e.preventDefault();
    }

    if (!isSaved) window.addEventListener('beforeunload', handleBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', handleBeforeUnload);
```
This is a bit simpler and easier to read, I think. `isEqual` would need importing from `lodash-es`, though. `savedChallenge` would come from the selector I mentioned above.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
The current project is only guaranteed to be first saved project if there's one saved challenge, so this can break if more than one project are saved.

To fix this, and DRY the codebase a little, you could create a `savedChallenge` selector, to get the current saved info for the current challenge. There are a few places in the codebase where we want the current saved challenge and they could all use it.

**Pull request closed without merging**